ath79: add support for COMFAST CF-E355AC v2

plaes opened the pull request:
# Hardware

COMFAST CF-E355AC v2 is a ceiling mount AP with PoE support,
based on Qualcomm/Atheros QCA9531 + QCA9886.

## Short specification:

- 1x 10/100 Mbps Ethernet, with PoE support (wan/eth1)
- 1x 10/100/1000 Mbps Ethernet, with PoE support (lan/eth0)
- 128MB of RAM (DDR2)
- 16 MB of FLASH
- 2T2R 2.4 GHz, 802.11b/g/n (wlan2g)
- 2T2R 5 GHz, 802.11ac/n/a, WAVE 2 (wlan5g)
- built-in 4x 3 dBi antennas
- output power (max): 500 mW (27 dBm)
- 1x RGB LED, 1x button
- separate watchdog chip (bottom of PCB?)

## Markings on PCB:

* R121QH_VER2.1 (silkscreen, bottom)
* CF-WA800 (sticker, top)

# Initial flashing instructions:

Original firmware is based on OpenWrt.

a) Use sysupgrade image directly in vendor GUI.

b) Or via tftp:

    ipaddr=192.168.1.1
    serverip=192.168.1.10
    bootfile="firmware.bin"

c) Or possibly via u-boot's `httpd` command.

## MAC addresses

MAC-address mapping follows differs somewhat from original firmware:

* eth1 (wan) is the lowest mac address (art @ 0x0)
* eth0 (lan) uses eth1 + 1 (art @ 0x1002)
* wlan2g uses eth1 + 2 (art @ 0x06)
* wlan5g uses eth1 + 3 (art @ 0x5006) **\*\*\*** **(original firmware is using eth1 + 10)**, no clue where it comes from.

Art dump (`hexdump /dev/mtd1 |grep ZZZZ`):

    0000000 ZZZZ XXXX XXX0 ZZZZ XXXX XXX2 ffff ffff
    0001000 0202 ZZZZ XXXX XXX1 0000 0000 0000 0000
    0005000 202f bd21 0101 ZZZZ XXXX XXX3 0000 2000

plaes committed:
ath79: add support for COMFAST CF-E355AC v2

COMFAST CF-E355AC v2 is a ceiling mount AP with PoE support,
based on Qualcomm/Atheros QCA9531 + QCA9886.

Short specification:

- 1x 10/100 Mbps Ethernet, with PoE support (wan/eth1)
- 1x 10/100/1000 Mbps Ethernet, with PoE support (lan/eth0)
- 128MB of RAM (DDR2)
- 16 MB of FLASH
- 2T2R 2.4 GHz, 802.11b/g/n (wlan2g)
- 2T2R 5 GHz, 802.11ac/n/a, WAVE 2 (wlan5g)
- built-in 4x 3 dBi antennas
- output power (max): 500 mW (27 dBm)
- 1x RGB LED, 1x button
- separate watchdog chip (bottom of PCB?)

Markings on PCB:

* R121QH_VER2.1 (silkscreen, bottom)
* CF-WA800 (sticker, top)

Initial flashing instructions:

Original firmware is based on OpenWrt.

a) Use sysupgrade image directly in vendor GUI.

b) Or via tftp:

  ipaddr=192.168.1.1
  serverip=192.168.1.10
  bootfile="firmware.bin"

c) Or possibly via u-boot's `httpd` command.

MAC-address mapping follows original firmware:

* eth1 (wan) is the lowest mac address (art @ 0x0)
* eth0 (lan) uses eth1 + 1 (art @ 0x1002)
* wlan2g uses eth1 + 2 (art @ 0x06)
* wlan5g uses eth1 + 10 (not present in art)
* unused MAC (eth1 + 3) (art @ 0x5006)

Art dump (`hexdump /dev/mtd1 |grep ZZZZ`):

  0000000 ZZZZ XXXX XXX0 ZZZZ XXXX XXX2 ffff ffff
  0001000 0202 ZZZZ XXXX XXX1 0000 0000 0000 0000
  0005000 202f bd21 0101 ZZZZ XXXX XXX3 0000 2000

Signed-off-by: Priit Laes <plaes@plaes.org>

adschm commented:
ar71xx -> ath79 in title ...

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e355ac-v2.dts:
```diff
@@ -0,0 +1,143 @@
+	 * Markings on PCB:
+	 *   R121QH_VER2.1 (silkscreen, bottom)
+	 *   CF-WA800 (sticker, top)
+	 */
```
Move that to commit message.

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e355ac-v2.dts:
```diff
@@ -0,0 +1,143 @@
+		};
+
+
+		led_wlan2g: wlan2g {
```
"led_wlan2g" DT label does not seem to be required? Same below ...

plaes replied:
Thanks, dropped in 25ba9d49

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e355ac-v2.dts:
```diff
@@ -0,0 +1,143 @@
+	status = "okay";
+};
+
+/* Is it needed? */
```
Well, do you have USB ports or other USB-connected components?

plaes replied:
Thanks, dropped this in the fixup comment.

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e355ac-v2.dts:
```diff
@@ -0,0 +1,143 @@
+	port@1 {
+		reg = <1>;
+		#trigger-source-cells = <0>;
+	};
```
port definition can be dropped if no USB LED is present.

plaes replied:
Thanks, dropped, because we don't have to enable USB at all.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -129,6 +129,11 @@ comfast,cf-e314n-v2)
+comfast,cf-e355-v2)
+	ucidef_set_led_netdev "lan" "LAN" "cf-e355ac-v2:green:lan" "eth0"
+	ucidef_set_led_wlan "wlan2g" "WLAN2G" "blue:wlan2g" "phy1tpt"
+	ucidef_set_led_wlan "wlan5g" "WLAN5G" "red:wlan5g" "phy0tpt"
```
WLAN definitions can be dropped here since you have the triggers DTS.

plaes replied:
Thanks, fixed in 244749b

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -129,6 +129,11 @@ comfast,cf-e314n-v2)
 	ucidef_set_led_rssi "rssihigh" "RSSIHIGH" "green:rssihigh" "wlan0" "76" "100"
 	;;
+comfast,cf-e355-v2)
+	ucidef_set_led_netdev "lan" "LAN" "cf-e355ac-v2:green:lan" "eth0"
```
remove model prefix.

plaes replied:
Thanks, fixed in 244749b

adschm commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -158,6 +158,11 @@ case "$FIRMWARE" in
 	;;
 "ath10k/pre-cal-pci-0000:00:00.0.bin")
 	case $board in
+	comfast,cf-e355ac-v2)
```
Sorting ...

plaes replied:
Thanks, fixed in 8767eca.

plaes commented:
I have gotten to the point where I can boot up the my device with the custom firmware. Although I have yet to validate whether I can push packets from wan, wlan2g and wlan5g interfaces, I ran into a bug? with the `ath10k_patch_mac` functionality, which is setting the mac address `12:34:56:78:90:12` on my test device.
I have validated the `$(macaddr_add $(mtd_get_mac_binary art 0x0) +10)` is returning proper mac address, so the issue lies somewhere in the `ath10k_patch_mac` function.

That issue seems also present with OpenWrt 19.07.5.

@adschm Should I squash commits and force-push my branch?

adschm commented:
> Should I squash commits and force-push my branch?

Yes, please.

plaes commented:
> > Should I squash commits and force-push my branch?
> 
> Yes, please.

Should be ready now.

There's still issue that MAC address for ath10k_pci device using `ath10k_patch_mac` does not work properly - it ends up as 12:34:56:78:90:12. But I also noticed this issue on the OpenWrt 19.07.5, r11257-5090152ae3 with the current ar71xx firmware.

s-2 commented:
> There's still issue that MAC address for ath10k_pci device using `ath10k_patch_mac` does not work properly - it ends up as 12:34:56:78:90:12. But I also noticed this issue on the OpenWrt 19.07.5, r11257-5090152ae3 with the current ar71xx firmware.

This looks like the QSDK default mac address placed in `art` (should be the one at `0x5002` here).

Beware that `11-ath10k-caldata` contains multiple switch cases for different types of ath10k chips; even though there are other comfast boards located nearby, I think your device belongs into the `"ath10k/cal-pci-0000:00:00.0.bin"` case statement above, rather than `"ath10k/pre-cal-pci-0000:00:00.0.bin"`. The parameter for caldata size also hints at this.

Honestly, I'm also not quite sure about how to successfully predict which is the correct case, so I usually end up injecting something like
`echo "~~~ $FIRMWARE ~~~" > /dev/ttyS0`
at the beginning of the script, so I could check the output on the serial console during script execution on boot 😇

s-2 commented on target/linux/ath79/dts/qca9531_comfast_cf-e355ac-v2.dts:
```diff
@@ -0,0 +1,136 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
```
can this be increased? usually we have `50000000` for the max value - considering this device already has an ath10k chip, also the flash chip is probably a newer model that supports even higher speeds, but with the CPU clock for these SoCs the maximum is usually 50 MHz.

plaes replied:
Ok, worked with 50MHz :)

s-2 commented on target/linux/ath79/dts/qca9531_comfast_cf-e355ac-v2.dts:
```diff
@@ -0,0 +1,136 @@
+			label = "green:lan";
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+		};
+
```
redundant empty line :)
also you might consider adding `default-state = "on";` to `led_lan`, in case there is a ~2 seconds gap where the led is turned off, right before the failsafe blinking pattern starts.

adschm replied:
Since this is a hijacked LAN LED, I'd rather not put the default-state here, so we have some chance to see the actual LED behavior from bootloader ...

plaes commented:
> > There's still issue that MAC address for ath10k_pci device using `ath10k_patch_mac` does not work properly - it ends up as 12:34:56:78:90:12. But I also noticed this issue on the OpenWrt 19.07.5, r11257-5090152ae3 with the current ar71xx firmware.
> 
> This looks like the QSDK default mac address placed in `art` (~should be the one at `0x5002` here~ ...wonder where it comes from, according to your art dump it is not at the default position).
> 
> Beware that `11-ath10k-caldata` contains multiple switch cases for different types of ath10k chips; even though there are other comfast boards located nearby, I think your device belongs into the `"ath10k/cal-pci-0000:00:00.0.bin"` case statement above, rather than `"ath10k/pre-cal-pci-0000:00:00.0.bin"`. The parameter for caldata size also hints at this.
> 
> Honestly, I'm also not quite sure about how to successfully predict which is the correct case, so I usually end up injecting something like
> `echo "~~~ $FIRMWARE ~~~" > /dev/ttyS0`
> at the beginning of the script, so I could check the output on the serial console during script execution on boot innocent

After deleting `/lib/firmware/ath10k/QCA9888/hw2.0/board-2.bin`, it ended up with correct MAC address.

Boot log with `board-2.bin` being present:
```
LOAD FW: ath10k/fwcfg-pci-0000:00:00.0.txt
LOAD FW: ath10k/fwcfg-pci-0000:00:00.0.txt
LOAD FW: ath10k/pre-cal-pci-0000:00:00.0.bin
LOAD FW: ath10k/QCA9888/hw2.0/ct-firmware-5.bin
LOAD FW: ath10k/QCA9888/hw2.0/ct-firmware-5.bin
LOAD FW: ath10k/QCA9888/hw2.0/ct-firmware-2.bin
LOAD FW: ath10k/QCA9888/hw2.0/ct-firmware-2.bin
LOAD FW: ath10k/QCA9888/hw2.0/firmware-6.bin
LOAD FW: ath10k/QCA9888/hw2.0/firmware-6.bin
ath10k_pci 0000:00:00.0: qca9888 hw2.0 target 0x01000000 chip_id 0x00000000 sub 0000:0000
ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
ath10k_pci 0000:00:00.0: firmware ver 10.4b-ct-9888-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 59e741e7
ath10k_pci 0000:00:00.0: board_file api 2 bmi_id 0:24 crc32 f228337a
ath10k_pci 0000:00:00.0: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
ath10k_pci 0000:00:00.0: msdu-desc: 2500  skid: 32
ath10k_pci 0000:00:00.0: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
ath10k_pci 0000:00:00.0: wmi print 'free: 114572 iram: 12644 sram: 29508'
ath10k_pci 0000:00:00.0: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-file max-sta 32 raw 0 hwcrypto 1
```

And boot log after deleting `board-2.bin`:
```
LOAD FW: ath10k/fwcfg-pci-0000:00:00.0.txt
LOAD FW: ath10k/fwcfg-pci-0000:00:00.0.txt
LOAD FW: ath10k/QCA9888/hw2.0/ct-firmware-5.bin
LOAD FW: ath10k/QCA9888/hw2.0/ct-firmware-5.bin
LOAD FW: ath10k/QCA9888/hw2.0/ct-firmware-2.bin
LOAD FW: ath10k/QCA9888/hw2.0/ct-firmware-2.bin
LOAD FW: ath10k/QCA9888/hw2.0/firmware-6.bin
LOAD FW: ath10k/QCA9888/hw2.0/firmware-6.bin
ath10k_pci 0000:00:00.0: qca9888 hw2.0 target 0x01000000 chip_id 0x00000000 sub 0000:0000
ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
ath10k_pci 0000:00:00.0: firmware ver 10.4b-ct-9888-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 59e741e7
LOAD FW: ath10k/QCA9888/hw2.0/board-2.bin
ath10k_pci 0000:00:00.0: board_file api 1 bmi_id 0:24 crc32 7e5f3941
LOAD FW: ath10k/QCA9888/hw2.0/board-2.bin
ath10k_pci 0000:00:00.0: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
ath10k_pci 0000:00:00.0: msdu-desc: 2500  skid: 32
ath10k_pci 0000:00:00.0: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
ath10k_pci 0000:00:00.0: wmi print 'free: 114572 iram: 12644 sram: 29508'
ath10k_pci 0000:00:00.0: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-file max-sta 32 raw 0 hwcrypto 1
```

Diff between these two is following (after skipping the `LOAD FW:` lines):
```
-ath10k_pci 0000:00:00.0: board_file api 2 bmi_id 0:24 crc32 f228337a
+ath10k_pci 0000:00:00.0: board_file api 1 bmi_id 0:24 crc32 7e5f3941
```

plaes commented:
@adschm, @s-2, I think I have fixed all the issues now (whitespace and wlan5g mac address) and I wouldn't touch the LED, as suggested by @adschm.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -14,6 +14,7 @@ case "$board" in
 alfa-network,ap121f|\
 alfa-network,ap121fe|\
 avm,fritz450e|\
+comfast,cf-e355-v2|\
```
"ac" missing. Testing would have shown that ...

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -530,6 +530,17 @@ define Device/comfast_cf-e314n-v2
+  DEVICE_MODEL := CF-E355AC
+  DEVICE_VARIANT := v2
+  DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca9888-ct \
+	  -swconfig -uboot-envtools
```
hanging indent with one tab only

adschm commented:
This still needs to be squashed properly.

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e355ac-v2.dts:
```diff
@@ -0,0 +1,135 @@
+
+	aliases {
+		serial0 = &uart;
+		label-mac-device = &eth1;
```
&eth1 is 0x0 +1, so probably eth0 on the running device. Am I correct and this is wrong?

adschm commented:
Please choose one of the two Signed-off-by lines and drop the other one.

adschm commented:
Ping.

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e355ac-v2.dts:
```diff
@@ -0,0 +1,135 @@
+
+	aliases {
+		serial0 = &uart;
+		label-mac-device = &eth1;
```
Please respond to my earlier comment. This is probably wrong, and the label MAC address should be that of WAN, i.e. 0x0 +0?

plaes replied:
No, eth0 is LAN, eth1 is WAN:
```
ip addr
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel master br-lan state UP qlen 1000
    link/ether 20:0d:b0:7a:e6:c1 brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP qlen 1000
    link/ether 20:0d:b0:7a:e6:c0 brd ff:ff:ff:ff:ff:ff
    inet 10.42.0.77/24 brd 10.42.0.255 scope global eth1
```

adschm replied:
eth0 is probably NOT &eth0, that's what I mean, since the DT names and runtime names are probably swapped!

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e355ac-v2.dts:
```diff
@@ -0,0 +1,135 @@
+
+&uart {
+	status = "okay";
+};
```
This block can be removed.

adschm commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -187,6 +187,13 @@ case "$FIRMWARE" in
 		;;
+	comfast,cf-e355ac-v2)
+		caldata_extract "art" 0x5000 0x844
+		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary art 0x0) +3)
```
This should be "10" instead of "+3" according to OEM firmware?

plaes replied:
Should we match the OEM or could we have nice sequential mac addresses:
```
2: eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc fq_codel state DOWN qlen 1000
    link/ether 20:0d:b0:7a:e6:c1 brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether 20:0d:b0:7a:e6:c0 brd ff:ff:ff:ff:ff:ff
4: wlan0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether 20:0d:b0:7a:e6:c2 brd ff:ff:ff:ff:ff:ff
5: wlan1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether 20:0d:b0:7a:e6:c3 brd ff:ff:ff:ff:ff:ff
```

adschm replied:
We should match the OEM so we have nice correct MAC addresses and nice collision prevention in networks.

adschm commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -187,6 +187,13 @@ case "$FIRMWARE" in
 		rm /lib/firmware/ath10k/QCA9888/hw2.0/board-2.bin
 		;;
+	comfast,cf-e355ac-v2)
+		caldata_extract "art" 0x5000 0x844
```
Are you really the only device just needing 0x844 here while all other need 0x2f20?

adschm reviewed (changes requested):
Almost pulled this, but there are a few issues left.

plaes commented:
Updated according to comments.
Changes:
* Dropped uart definition from devicetree
* Moved device definition in 11-ath10k-caldata to "common" section (with this change we end up mac address of WAN + 2, which differs from the OEM definition of WAN + 10
* Had to increase the 2.4G wifi mac address by one (WAN + 3 now, otherwise it clashed with wlan5g)

Interfaces:
```
eth0: 20:0d:b0:7a:e6:c1 (LAN)
eth1: 20:0d:b0:7a:e6:c0 (WAN)
wlan0: 20:0d:b0:7a:e6:c2 (5G)
wlan1: 20:0d:b0:7a:e6:c3 (2.4G)
```

adschm commented:
> Had to increase the 2.4G wifi mac address by one (WAN + 3 now, otherwise it clashed with wlan5g)
> Interfaces:
> 
> eth0: 20:0d:b0:7a:e6:c1 (LAN)
> eth1: 20:0d:b0:7a:e6:c0 (WAN)
> wlan0: 20:0d:b0:7a:e6:c2 (5G)
> wlan1: 20:0d:b0:7a:e6:c3 (2.4G)

As I said multiple times already, please do it like the vendor does it. Is it really that hard?

plaes commented:
Closing in favor of #16556
